go/cmd/exporter/worker.go

```diff
@@ -0,0 +1,266 @@
+package main
+
+import (
+	"archive/zip"
+	"bytes"
+	"cmp"
+	"context"
+	"encoding/csv"
+	"encoding/json"
+	"io"
+	"log/slog"
+	"maps"
+	"path/filepath"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/osv.dev/go/logger"
+	"github.com/ossf/osv-schema/bindings/go/osvschema"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+const (
+	gitEcosystem        = "GIT"
+	vanirVulnsFilename  = "osv_git.json"
+	allZipFilename      = "all.zip"
+	modifiedCSVFilename = "modified_id.csv"
+	ecosystemsFilename  = "ecosystems.txt"
+)
+
+// ecosystemWorker processes vulnerabilities for a single ecosystem.
+type ecosystemWorker struct {
+	ecosystem string
+	inCh      chan *osvschema.Vulnerability
+}
+
+// newEcosystemWorker creates and starts a new ecosystemWorker.
+func newEcosystemWorker(ctx context.Context, ecosystem string, outCh chan<- writeMsg, wg *sync.WaitGroup) *ecosystemWorker {
+	ch := make(chan *osvschema.Vulnerability)
+	worker := &ecosystemWorker{
+		ecosystem: ecosystem,
+		inCh:      ch,
+	}
+	wg.Add(1)
+	go worker.run(ctx, outCh, wg)
+
+	return worker
+}
+
+// vulnData holds the ID and marshalled JSON data for a vulnerability.
+type vulnData struct {
+	id   string
+	data []byte
+}
+
+var protoMarshaller = protojson.MarshalOptions{
+	UseProtoNames: true, // TODO(michaelkedar): https://github.com/ossf/osv-schema/pull/442
+}
+
+// run is the main loop for the ecosystemWorker. It receives vulnerabilities,
+// aggregates them, and upon completion, writes out the ecosystem-specific
+// zip, csv, and (for GIT) vanir files.
+func (w *ecosystemWorker) run(ctx context.Context, outCh chan<- writeMsg, wg *sync.WaitGroup) {
+	defer wg.Done()
+	logger.Info("new ecosystem worker started", slog.String("ecosystem", w.ecosystem))
+	var allVulns []vulnData
+	var csvData [][]string
+	var vanirVulns []vulnData
+	for {
+		var v *osvschema.Vulnerability
+		var ok bool
+
+		// Wait to receive a vulnerability, or be cancelled.
+		select {
+		case v, ok = <-w.inCh:
+			if !ok {
+				logger.Info("All vulnerabilities processed", slog.String("ecosystem", w.ecosystem))
+				writeModifiedIDCSV(ctx, filepath.Join(w.ecosystem, modifiedCSVFilename), csvData, outCh)
+				writeZIP(ctx, filepath.Join(w.ecosystem, allZipFilename), allVulns, outCh)
+				if w.ecosystem == gitEcosystem {
+					writeVanir(ctx, vanirVulns, outCh)
+				}
+				logger.Info("ecosystem worker finished processing", slog.String("ecosystem", w.ecosystem))
+
```

Comment:
Break here and put the finalising code at the bottom of the func (same with the other loops)